### ISSUE TITLE
Deploying Kubeflow to Azure with DNS and TLS termination 

### ISSUE THREAD

**Issue by lauradang** [kubeflow/kubeflow#5976]
/kind feature

Similarly to the [documentation regarding deploying Kubeflow to IBM Cloud with DNS and TLS termination](https://www.kubeflow.org/docs/distributions/ibm/deploy/authentication/#exposing-the-kubeflow-dashboard-with-dns-and-tls-termination), it would be great to have similar documentation for deploying Kubeflow to Azure with DNS and TLS termination. Right now, it only shows us how to deploy with a self-signed certificate, which results in a untrusted certificate connection situation. I'm trying to do it myself without knowing much Azure, and it's proven to be quite difficult. 

If anyone has their own solution, I would really appreciate hearing it.

Thanks!

**Comment by DavidSpek** [comment]
I recently ran @pablofiumara through the process of creating a proper certificate on Azure and setting up the Istio Gateway to use that certificate. I haven't had the time yet to write more docs about this, but maybe he can help you out with this a bit.

**Comment by pablofiumara** [comment]
I will post some guidelines in the next days to have SSL enabled on Azure. Meanwhile I suggest reading readme file here and its code: https://github.com/argoflow/argoflow-azure

**Comment by pablofiumara** [comment]
@lauradang Execute this commands

1) 
az group create -n resourcetest3 -l westus
2) 
az aks create -g resourcetest3 -n clustertest3 -s Standard_D4s_v3 -c 4 -l westus --generate-ssh-keys --kubernetes-version 1.19.11
3) 
az aks get-credentials -n clustertest3 -g resourcetest3 --overwrite-existing

4) Buy a domain (not one owned by Azure) and you assign it to resourcetest2. You can do it here (App Service Domains): https://portal.azure.com/#blade/HubsExtension/BrowseResource/resourceType/Microsoft.DomainRegistration%2Fdomains
5) 
Fork https://github.com/argoflow/argoflow-azure
6) 
git clone yourForkUrl (something like this: https://github.com/yourUser/argoflow-azure  )
7) 
cd argoflow-azure/


8) 
AZURE_CERT_MANAGER_SP_NAME=certmanagerserviceprincipalresourcetest3
9) 
AZURE_CERT_MANAGER_DNS_RESOURCE_GROUP=resourcetest3
10) 
AZURE_CERT_MANAGER_DNS_NAME=yourdomain.com
11) 
DNS_SP=$(az ad sp create-for-rbac --name $AZURE_CERT_MANAGER_SP_NAME)
12) 
AZURE_CERT_MANAGER_SP_APP_ID=$(echo $DNS_SP | jq -r '.appId')
13) 
AZURE_CERT_MANAGER_SP_PASSWORD=$(echo $DNS_SP | jq -r '.password')

14) 
az role assignment delete --assignee $AZURE_CERT_MANAGER_SP_APP_ID --role Contributor
15) 
DNS_ID=$(az network dns zone show --name $AZURE_CERT_MANAGER_DNS_NAME --resource-group $AZURE_CERT_MANAGER_DNS_RESOURCE_GROUP --query "id" --output tsv)
16) 
az role assignment create --assignee $AZURE_CERT_MANAGER_SP_APP_ID --role "DNS Zone Contributor" --scope $DNS_ID
17) 
az role assignment list --assignee $AZURE_CERT_MANAGER_SP_APP_ID

18) 
kubectl create namespace cert-manager
19) 
kubectl create secret -n cert-manager generic azuredns-config --from-literal=CLIENT_SECRET=$AZURE_CERT_MANAGER_SP_PASSWORD
20) 
echo "Principal: $AZURE_CERT_MANAGER_SP_APP_ID"    
21) 
echo "Password: $AZURE_CERT_MANAGER_SP_PASSWORD"   

This will create an app here: https://portal.azure.com/#blade/Microsoft_AAD_IAM/ActiveDirectoryMenuBlade/RegisteredApps  Reference: https://docs.cert-manager.io/en/release-0.11/tasks/issuers/setup-acme/dns01/azuredns.html



22a) 
vim distribution/cert-manager/lets-encrypt-dns-01/patches/cluster-issuer.yaml

22b) Paste this:

```
apiVersion: cert-manager.io/v1
kind: ClusterIssuer
metadata:
  name: gateways-issuer
spec:  
  selfSigned:
    $patch: delete
  acme:
    email: yourEmail@gmail.com
    server: https://acme-v02.api.letsencrypt.org/directory
    privateKeySecretRef:
          name: gateways-issuer-account-key
    solvers:
    - dns01:
        azureDNS:
          subscriptionID: yourSubscriptionId
          resourceGroupName: resourcetest3
          hostedZoneName: yourdomain.com
          # Azure Cloud Environment, default to AzurePublicCloud
          environment: AzurePublicCloud
          clientID: yourClientIdGetItFromAboveAZURE_CERT_MANAGER_SP_APP_ID
          clientSecretSecretRef:
            name: azuredns-config
            key: CLIENT_SECRET
          tenantID: yourTenantId



```
yourSubscriptionId: Get it from here https://portal.azure.com/#blade/Microsoft_Azure_Billing/SubscriptionsBlade
tenantId: Enter into your client and copy your tenant id https://portal.azure.com/#blade/Microsoft_AAD_IAM/ActiveDirectoryMenuBlade/RegisteredApps





22c) 
vim examples/setup.conf 

 (suppose your email is user@gmail.com  ) and write the following:

<<__git_repo.url__>>=yourForkUrl
<<__git_repo.target_revision__>>=main
<<__domain__>>=yourdomain.com
<<__subdomain_dashboard__>>=kubeflow
<<__subdomain_serving__>>=serving
<<__subdomain_argocd__>>=argocd
<<__subdomain_auth__>>=auth
<<__subdomain_grafana__>>=grafana
<<__subdomain_kiali__>>=kiali
<<__subdomain_kubecost__>>=kubecost
<<__oidc.issuer__>>=https://auth.my-cluster.my-domain.com
<<__cert_manager.email_user__>>=user
<<__cert_manager.email_domain__>>=gmail.com
<<__cert_manager.server__>>=https://acme-v02.api.letsencrypt.org/directory
<<__cloudflare.email__>>=info@my-domain.com


23) 
vim distribution/kustomization.yaml

Cert manager section should look like this:

```
  # - argocd-applications/cert-manager-self-signing.yaml
   -  argocd-applications/cert-manager-dns-01.yaml

```
24) 
wget https://github.com/bitnami-labs/sealed-secrets/releases/download/v0.16.0/kubeseal-linux-amd64 -O kubeseal

25) 
sudo install -m 755 kubeseal /usr/local/bin/kubeseal
26) 
pip install passlib==1.7.4

27)  Install kustomize 3.2.0. Tutorial: https://kubectl.docs.kubernetes.io/installation/kustomize/source/  

28) 
cd distribution/
 
29)
 ~/go/bin/kustomize build argocd/base/ | kubectl apply -f -

30) 
kubectl apply -f argocd-applications/sealed-secrets.yaml
31)
cd ..
32)
./setup_repo.sh ./examples/setup.conf

When asked a question and you need to answer yes or no, press 2 (no) and then enter

33) 
git add .
34)
git commit -m "installation"
35)
git push -u origin main
36)
kubectl apply -f distribution/kubeflow.yaml

37) 
kubectl get pods --all-namespaces 

After one minute, many pods will appear. Wait until every pod (except kfserving-controller-manager) have Running status

38)
 kubectl get svc -n istio-system istio-ingressgateway -o jsonpath='{.status.loadBalancer.ingress[0]}'

39) Get the ip from previous step, choose your dns zone here: https://portal.azure.com/#blade/HubsExtension/BrowseResource/resourceType/Microsoft.Network%2FdnsZones

40) Click on record set

name is kubeflow
Type the ip
Press OK

41) Add a record set for auth and for serving as well (ip is the same)

42) 
vim distribution/oidc-auth/overlays/keycloak/keycloak.yaml

Delete these lines:

      - name: metrics.enabled
        value: "true"
      - name: metrics.serviceMonitor.enabled
        value: "true"
      - name: metrics.serviceMonitor.namespace
        value: monitoring
        
43)  
git add . && git commit -m "deleted metrics" && git push origin main       

44)
kubectl apply -f distribution/oidc-auth/overlays/keycloak/keycloak.yaml


47)

kubectl edit mutatingwebhookconfiguration istio-sidecar-injector -n kubeflow

48) Add this below metadata.annotations:

   admissions.enforcer/disabled: "true"
   
49) Under namespaceSelector.matchExpressions delete lines

    - key: control-plane
      operator: DoesNotExist
      
There are 4 places where you need to delete those lines

50) 

kubectl delete pods --all -n kubeflow

51)
kubectl get pods -n kubeflow

Wait until every pod is on Running state except kfserving-controller-manager



Login to kubeflow.yourdomain

**Comment by lauradang** [comment]
@pablofiumara Thank you so much for this! Since we do not have a paid domain, I want to use a public IP address and a DNS label as the domain as shown here: https://docs.microsoft.com/en-us/azure/aks/static-ip. I would edit this service: `istio-ingressgateway`'s load balancer IP and annotations. This annotation would be my domain. Would that also work? If so, are there any differences to this config you laid out?

**Comment by lauradang** [comment]
@pablofiumara @DavidSpek Thank you both of you! Since we do not have a paid domain, I want to use a public IP address and a DNS label as the domain as shown here: https://docs.microsoft.com/en-us/azure/aks/static-ip. I would edit this service: `istio-ingressgateway`'s load balancer IP and annotations. Then my domain would be:
`<my_dns_label>.<location>.cloudapp.azure.com`

 Would that also work? If so, are there any differences to this config you laid out?

For reference, I tried to do this (change load balancer IP and annotations as said in the Microsoft documentation), and I also added kubeflow, auth, and serving A records to a DNS Zone I created. When I tried to access `kubeflow.<my_dns_label>.<location>.cloudapp.azure.com`, this error came up on my browser:
`kubeflow.<my_dns_label>.<location>.cloudapp.azure.com’s server IP address could not be found.`

I also tried with `auth.` and `serving.` and the same errors come up. However, when accessing `<my_dns_label>.<location>.cloudapp.azure.com` (no `kubeflow` prefix), I get an error that says:
`RBAC: access denied`

**Comment by pablofiumara** [comment]
@lauradang I tried using one of those domains owned by Azure and it didn't work. Perhaps it didn't work because I was not generating the SSL certificate correctly. I suggest you try the steps I sent you using domain *cloudapp.azure.com and see if it works

**Comment by lauradang** [comment]
@pablofiumara Unfortunately, that doesn't change anything - the same error comes up when I use the domain *cloudapp.azure.com. Since this does not work, are these the only ways to deploy Kubeflow with SSL enabled **in Azure**:
- Buy a domain and follow the steps you have laid out
- Generate a certificate yourself and get a CA to sign off on it (Digicert, Globalsign, etc.), create a secret of this certificate, and point the istio-ingressway service to this secret

In other words, there is no way to use the DNS label generated by Azure Public IP Address and enable SSL with Kubeflow?

**Comment by lauradang** [comment]
@pablofiumara 
I've also tried to use `cert-manager` DNS01 to generate a signed certificate with one of the domains owned by Azure before, but it didn't work for me either because the DNS was failing to propagate. 

Nonetheless, unfortunately, the same error comes up when I use the domain *cloudapp.azure.com. Since this does not work, are these the only ways to deploy Kubeflow with SSL enabled **in Azure**:
- Buy a domain and follow the steps you have laid out
- Generate a certificate yourself and get a CA to sign off on it (Digicert, Globalsign, etc.), create a secret of this certificate, and point the istio-ingressway service to this secret

In other words, is there no way to use the DNS label generated by Azure Public IP Address and enable SSL with Kubeflow?

**Comment by lauradang** [comment]
@pablofiumara 
I've also tried to use `cert-manager` DNS01 to generate a signed certificate with one of the domains owned by Azure before, but it didn't work for me either because the DNS was failing to propagate. 

Nonetheless, unfortunately, the same error comes up when I use the domain *cloudapp.azure.com. Since this does not work, are these the only ways to deploy Kubeflow with SSL enabled **in Azure**:
- Buy a domain and follow the steps you have laid out
- Generate a certificate yourself and get a CA to sign off on it (Digicert, Globalsign, etc.), create a secret of this certificate, and point the istio-ingressway service to this secret

In other words, is there no way to use the DNS label from Azure Public IP Address to enable SSL with Kubeflow?

**Comment by pablofiumara** [comment]
@lauradang I was able to enable SSL using one domain owned by Azure. However, I didn't manage to install Kubeflow on that cluster. I followed this tutorial: [https://docs.microsoft.com/en-us/azure/aks/ingress-static-ip](https://docs.microsoft.com/en-us/azure/aks/ingress-static-ip). I couldn't go from that setup to a Kubeflow installation. The cause of the issue might be related to what you mentioned: Kubeflow uses Istio Gateway, Microsoft tutorial uses Ingress resource

Regarding generating a certificate yourself and then creating a secret, I tried that and problems appeared when I tried to point the istio-ingressgateway to that secret. I generated the certificate following these steps: https://github.com/mspnp/letsencrypt-pip-cert-generation

**Comment by lauradang** [comment]
@pablofiumara 
I used this link: https://github.com/mspnp/letsencrypt-pip-cert-generation to generate the certificate. Then I actually had to manually go in and edit the `istio-ingressgateway-certs` secret since pointing the `Gateway` resource to the certificate secret using `credentialName` wasn't working for me either (Similarly to what was done when deploying Kubeflow to IBM Cloud: https://ai-at-scale.canadacentral.cloudapp.azure.com). So in the end, I now have Kubeflow deployed on Azure with full auth and SSL-enabled. Thanks for all the help!

**Comment by lauradang** [comment]
@pablofiumara 
I used this link: https://github.com/mspnp/letsencrypt-pip-cert-generation to generate the certificate. Then I actually had to manually go in and edit the `istio-ingressgateway-certs` secret since pointing the `Gateway` resource to the certificate secret using `credentialName` wasn't working for me either (Similarly to what was done when deploying Kubeflow to IBM Cloud: https://www.kubeflow.org/docs/distributions/ibm/deploy/authentication/#exposing-the-kubeflow-dashboard-with-dns-and-tls-termination). So in the end, I now have Kubeflow deployed on Azure with full auth and SSL-enabled. Thanks for all the help!
